source/snapshot/fetch_worker.go

```diff
@@ -0,0 +1,393 @@
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/conduitio/conduit-connector-postgres/source/position"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/google/uuid"
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgxpool"
+)
+
+const defaultFetchSize = 50000
+
+var supportedKeyTypes = []string{
+	"smallint", "integer", "bigint",
+	"decimal", "numeric", "real", "double",
+}
+
+type FetchConfig struct {
+	Table     string
+	Key       string
+	Snapshot  string
+	FetchSize int
+	Position  position.Position
+}
+
+var (
+	errTableRequired  = errors.New("table name is required")
+	errKeyRequired    = errors.New("table key required")
+	errInvalidCDCType = errors.New("invalid position type CDC")
+)
+
+func (c FetchConfig) Validate() error {
+	var errs []error
+
+	if c.Table == "" {
+		errs = append(errs, errTableRequired)
+	}
+
+	if c.Key == "" {
+		errs = append(errs, errKeyRequired)
+	}
+
+	switch c.Position.Type {
+	case position.TypeSnapshot, position.TypeInitial:
+	default:
+		errs = append(errs, errInvalidCDCType)
+	}
+
+	if len(errs) != 0 {
+		return errors.Join(errs...)
+	}
+
+	return nil
+}
+
+type FetchWorker struct {
+	conf FetchConfig
+	db   *pgxpool.Pool
+	out  chan<- sdk.Record
+
+	snapshotEnd int64
+	lastRead    int64
+	cursorName  string
+}
+
+func NewFetchWorker(db *pgxpool.Pool, out chan<- sdk.Record, c FetchConfig) *FetchWorker {
+	f := &FetchWorker{
+		conf:       c,
+		db:         db,
+		out:        out,
+		cursorName: "fetcher_" + strings.ReplaceAll(uuid.NewString(), "-", ""),
+	}
+
+	if f.conf.FetchSize == 0 {
+		f.conf.FetchSize = defaultFetchSize
+	}
+
+	if c.Position.Type == position.TypeInitial || c.Position.Snapshot == nil {
+		return f
+	}
+
+	if t, ok := c.Position.Snapshot[c.Table]; ok {
+		f.snapshotEnd = t.SnapshotEnd
+		f.lastRead = t.LastRead
+	}
+
+	return f
+}
+
+// Validate will ensure the config is correctt.
+// * Table and keys exist
+// * Key is a primary key
+func (f *FetchWorker) Validate(ctx context.Context) error {
+	if err := f.conf.Validate(); err != nil {
+		return fmt.Errorf("failed to validate config: %w", err)
+	}
+
+	tx, err := f.db.Begin(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to start tx for validation: %w", err)
+	}
+	defer func() {
+		if err := tx.Rollback(ctx); err != nil {
+			sdk.Logger(ctx).Warn().
+				Err(err).
+				Msgf("error on validation tx rollback for %q", f.cursorName)
+		}
+	}()
+
+	if err := f.validateTable(ctx, f.conf.Table, tx); err != nil {
+		return fmt.Errorf("failed to validate table: %w", err)
+	}
+
+	if err := f.validateKey(ctx, f.conf.Table, f.conf.Key, tx); err != nil {
+		return fmt.Errorf("failed to validate key: %w", err)
+	}
+
+	return nil
+}
+
+func (f *FetchWorker) Run(ctx context.Context) error {
+	start := time.Now().UTC()
+
+	tx, err := f.db.BeginTx(ctx, pgx.TxOptions{
+		IsoLevel:   pgx.RepeatableRead,
+		AccessMode: pgx.ReadOnly,
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := tx.Rollback(ctx); err != nil {
```

Comment:
Just thinking out loud - is a rollback appropriate here? Would it make sense to rather commit? It's not like we are changing anything. Or rollback in case we are returning an error and commit otherwise. Does it make a difference?

Reply:
The tx will be discarded anyway, so it doesn't matter. We can commit, I just don't see a reason why.
Cursors require an active tx. Since we are not changing anything didn't make sense. I will double check

Reply:
Don't waste time on this, I was just wondering if there's a case to be made for one or the other. Probably it won't make a difference either way.

Reply:
Had another look at the [docs](https://www.postgresql.org/docs/current/sql-declare.html)

TL;DR it only matters if `WITH HOLD` is specified which will persist the cursor when the transaction completes,
in which case it makes sense to commit rather than rollback.

As an aside, in some situations `WITH HOLD` can be used with the following considerations.

1. Declare cursor with hold and it exists outside of the TX 
2. FETCH in parallel from multiple connections, but this can screw up ordering if one is fetching faster.
3. Cursor will need to be closed explicitly in this case.

One beneficial use to using this option is if we don't want to maintain the transaction open and this may become very reasonable if locks are causing contention, in which case the flow will be something like this:

1. Generate consistent cursor name
2. Declare cursor with hold
3. Fetch in a single connection 
4. Close cursor when done

Resuming becomes more complicated, because we will either need to close and redeclare the cursor (based in the name provided in 1), or figure out current position of cursor and seek back, but this will require the `SCROLL`  option.

I'm happy to leave it as is, until we need to put more work.